src/Features/CSharp/Portable/InternalUtilities/CSharpOperationAnalysisHelpers.cs

```diff
@@ -0,0 +1,941 @@
+            };
+
+            bool TryGetBoundVariableForIsPattern(IPatternOperation isPattern, [NotNullWhen(true)] out ISymbol? boundVariable)
+            {
+                boundVariable = null;
+                // got to the leftmost condition and it is an "is" pattern
+                if (!successRequirement)
+                {
+                    // we could be in an "expect false for successful equality" condition
+                    // and so we would want the pattern to be an "is not" pattern instead of an "is" pattern
+                    if (isPattern is INegatedPatternOperation negatedPattern)
+                    {
+                        isPattern = negatedPattern.Pattern;
+                    }
+                    else
+                    {
+                        // if we don't see "is not" then the pattern sequence is incorrect
+                        return false;
+                    }
+                }
+
+                if (isPattern is IDeclarationPatternOperation
+                    {
+                        DeclaredSymbol: ISymbol otherVar,
+                        MatchedType: INamedTypeSymbol matchedType,
+                    } &&
+                    matchedType.Equals(currentObject.GetSymbolType()) &&
+                    // found the correct binding, add any members we equate in the rest of the binary condition
+                    // if we were in a binary condition at all, and signal failure if any condition is bad
+                    additionalConditions.All(otherCondition => TryAddEqualizedFieldsForCondition(
+                        otherCondition, successRequirement, currentObject, otherVar, builder)))
+                {
+                    boundVariable = otherVar;
+                    return true;
+                }
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Match a list of statements and add members that are compared
+        /// </summary>
+        /// <param name="statementsToCheck">operations which should compare members</param>
+        /// <param name="otherC">non-this comparison of the type we're equating</param>
+        /// <param name="type">the this symbol</param>
+        /// <param name="builder">builder for property/field comparisons we encounter</param>
+        /// <returns>whether every statment was one of the expected forms</returns>
+        private static bool TryAddEqualizedFieldsForStatements(
+            IEnumerable<IOperation> statementsToCheck,
+            ISymbol otherC,
+            INamedTypeSymbol type,
+            ArrayBuilder<IFieldSymbol> builder)
+        {
+            while (!statementsToCheck.IsEmpty())
+            {
+                // iterate. We don't use a for loop because we could potentially
+                // be changing this list based on what statements we see
+                var statement = statementsToCheck.First();
+                statementsToCheck = statementsToCheck.Skip(1);
```

Comment:
Hmm, this means we're potentially creating IEnumerables around IEnumerables here since this in a loop; can we just switch this to use recursion and then this becomes a regular loop?